src/renderer/vt/Xterm256Engine.cpp

```diff
@@ -9,102 +9,79 @@ using namespace Microsoft::Console::Render;
-        const bool flagSet = WI_AreAllFlagsSet(extendedAttrs, attr);
-        const bool lastState = WI_AreAllFlagsSet(_lastExtendedAttrsState, attr);
-        if (flagSet != lastState)
-        {
-            if (flagSet)
-            {
-                RETURN_IF_FAILED(beginFn(this));
-            }
-            else
-            {
-                RETURN_IF_FAILED(endFn(this));
-            }
-            WI_ToggleAllFlags(_lastExtendedAttrsState, attr);
-        }
-        return S_OK;
-    };
+    if (textAttributes.IsBold() != _lastTextAttributes.IsBold())
+    {
+        RETURN_IF_FAILED(_SetBold(textAttributes.IsBold()));
+        _lastTextAttributes.SetBold(textAttributes.IsBold());
+    }
+
+    if (textAttributes.IsUnderlined() != _lastTextAttributes.IsUnderlined())
+    {
+        RETURN_IF_FAILED(_SetUnderline(textAttributes.IsUnderlined()));
+        _lastTextAttributes.SetUnderline(textAttributes.IsUnderlined());
+    }
+
+    if (textAttributes.IsItalic() != _lastTextAttributes.IsItalic())
+    {
+        RETURN_IF_FAILED(_SetItalics(textAttributes.IsItalic()));
+        _lastTextAttributes.SetItalics(textAttributes.IsItalic());
+    }
 
-    auto hr = updateFlagAndState(ExtendedAttributes::Italics,
-                                 &Xterm256Engine::_BeginItalics,
-                                 &Xterm256Engine::_EndItalics);
-    RETURN_IF_FAILED(hr);
+    if (textAttributes.IsBlinking() != _lastTextAttributes.IsBlinking())
+    {
+        RETURN_IF_FAILED(_SetBlinking(textAttributes.IsBlinking()));
+        _lastTextAttributes.SetBlinking(textAttributes.IsBlinking());
+    }
 
-    hr = updateFlagAndState(ExtendedAttributes::Blinking,
-                            &Xterm256Engine::_BeginBlink,
-                            &Xterm256Engine::_EndBlink);
-    RETURN_IF_FAILED(hr);
+    if (textAttributes.IsInvisible() != _lastTextAttributes.IsInvisible())
+    {
+        RETURN_IF_FAILED(_SetInvisible(textAttributes.IsInvisible()));
+        _lastTextAttributes.SetInvisible(textAttributes.IsInvisible());
+    }
 
-    hr = updateFlagAndState(ExtendedAttributes::Invisible,
-                            &Xterm256Engine::_BeginInvisible,
-                            &Xterm256Engine::_EndInvisible);
-    RETURN_IF_FAILED(hr);
+    if (textAttributes.IsCrossedOut() != _lastTextAttributes.IsCrossedOut())
+    {
+        RETURN_IF_FAILED(_SetCrossedOut(textAttributes.IsCrossedOut()));
+        _lastTextAttributes.SetCrossedOut(textAttributes.IsCrossedOut());
+    }
 
-    hr = updateFlagAndState(ExtendedAttributes::CrossedOut,
-                            &Xterm256Engine::_BeginCrossedOut,
-                            &Xterm256Engine::_EndCrossedOut);
-    RETURN_IF_FAILED(hr);
+    if (textAttributes.IsReverseVideo() != _lastTextAttributes.IsReverseVideo())
+    {
+        RETURN_IF_FAILED(_SetReverseVideo(textAttributes.IsReverseVideo()));
+        _lastTextAttributes.SetReverseVideo(textAttributes.IsReverseVideo());

```

Comment:
This is one of those rare cases where I might _prefer_ a macro. This opinion is not likely held by the team (but I bet it could be socialized). What do you think? It's mechanical substring replacement, admittedly less debuggable, but repetition breeds potential for errors :smile:

```c++
#define UPDATE_STATE(stateName) \
 if(textAttributes.Is##stateName##() != _lastTextAttributes.Is##stateName##()) { \
  R_I_F(_Set##stateName##(...

UPDATE_STATE(CrossedOut)
UPDATE_STATE(Invisible)

#undef ...
```

Reply:
Personally I'm not a fan of macros, since I find it makes the code harder to understand. Like there's no indication what properties are being referenced or updated in that `UPDATE_STATE` call. But if you guys prefer that sort of thing, I don't mind changing it - I can see the argument for minimizing the repetition.

However, I should add that in the longer term I would have proposed rewriting much of that code anyway, to bundle the SGR sequences (e.g. instead of sending `\e[m\e[37m\e[1m\e[7m`, we should really be sending something like `\e[0;37;1;7m`). And in that case, the code would look completely different, and could well be one line per attribute anyway.

Reply:
I'm not terribly bothered by it. If in the long term we'll evolve towards attribute bundling, I'm down for leaving this expanded to make it painful/obvious.

Reply:
I vote no macro and for James's longer term plan of sending one SGR with semicolons that makes this mostly obsolete.